src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNode.cs

```diff
@@ -0,0 +1,408 @@
+            {
+                Property = property;
+                Index = index;
+            }
+
+            protected override Test WithInputCore(Evaluation? newInput) => new IndexEvaluation(newInput, Property, Index, Syntax);
+        }
+
+        internal abstract class MemberEvaluation : Evaluation
+        {
+            protected MemberEvaluation(Evaluation? input, SyntaxNode? syntax) : base(input, syntax)
+            {
+            }
+
+            public abstract ISymbol Symbol { get; }
+        }
+
+        internal sealed class FieldEvaluation : MemberEvaluation
+        {
+            public readonly IFieldSymbol Field;
+            public override ISymbol Symbol => Field;
+            public FieldEvaluation(Evaluation? input, IFieldSymbol field, SyntaxNode? syntax = null) : base(input, syntax) => Field = field.CorrespondingTupleField ?? field;
+            protected override Test WithInputCore(Evaluation? newInput) => new FieldEvaluation(newInput, Field, Syntax);
+        }
+
+        internal sealed class PropertyEvaluation : MemberEvaluation
+        {
+            public readonly IPropertySymbol Property;
+            public override ISymbol Symbol => Property;
+            public PropertyEvaluation(Evaluation? input, IPropertySymbol property, SyntaxNode? syntax = null) : base(input, syntax) => Property = property;
+            protected override Test WithInputCore(Evaluation? newInput) => new PropertyEvaluation(newInput, Property, Syntax);
+        }
+
+        internal sealed class DeconstructEvaluation : Evaluation
+        {
+            public readonly IMethodSymbol DeconstructMethod;
+            public DeconstructEvaluation(Evaluation? input, IMethodSymbol deconstructMethod) : base(input) => DeconstructMethod = deconstructMethod;
+            protected override Test WithInputCore(Evaluation? newInput) => new DeconstructEvaluation(newInput, DeconstructMethod);
+        }
+
+        internal sealed class OutVariableEvaluation : Evaluation
+        {
+            public readonly int Index;
+            public OutVariableEvaluation(Evaluation? input, int index) : base(input)
+            {
+                Debug.Assert(input is DeconstructEvaluation);
+                Index = index;
+            }
+
+            protected override Test WithInputCore(Evaluation? newInput) => new OutVariableEvaluation(newInput, Index);
+        }
+
+        internal sealed class OperationEvaluation : Evaluation
+        {
+            public OperationEvaluation(IOperation operation) : base(input: null, operation.Syntax) { }
+            protected override Test WithInputCore(Evaluation? newInput) => throw ExceptionUtilities.Unreachable;
+        }
+
+#if DEBUG
+        internal string Dump()
+        {
+            return DumpNode(this);
+
+            static string DumpNode(AnalyzedNode node, int depth = 0)
+            {
+                return node switch
```

Comment:
i feel like just using abstract+inheritance actually works much more nicely for your node structure.

Reply:
Same thing for Equals/GetHashCode. I believe this is more simple and coherent. This one is debug-only anyways so I don't think it's important perf-wise.